sim/riscv/sim-main.c

```diff
@@ -1653,23 +2124,20 @@ void step_once (SIM_CPU *cpu)
   unsigned int len;
   sim_cia pc = cpu->pc;
   const struct riscv_opcode *op;
+  int xlen = RISCV_XLEN (cpu);
 
   if (TRACE_ANY_P (cpu))
     trace_prefix (sd, cpu, NULL_CIA, pc, TRACE_LINENUM_P (cpu),
 		  NULL, 0, " "); /* Use a space for gcc warnings.  */
 
   iw = sim_core_read_aligned_2 (cpu, pc, exec_map, pc);
 
-  /* Reject non-32-bit opcodes first.  */
   len = riscv_insn_length (iw);
-  if (len != 4)
-    {
-      sim_io_printf (sd, "sim: bad insn len %#x @ %#"PRIxTA": %#"PRIxTW"\n",
-		     len, pc, iw);
-      sim_engine_halt (sd, cpu, NULL, pc, sim_signalled, SIM_SIGILL);
-    }
 
-  iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc + 2) << 16);
+  if (len == 4)
+    iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc + 2) << 16);
+  else
+    iw |= ((unsigned_word)sim_core_read_aligned_2 (cpu, pc, exec_map, pc));
```

Comment:
This looks redundant.  For a 2-byte instruction, we already read two bytes.  No need to re-read the same 2 bytes again.